### PR TITLE
perl 5.26.0

### PR DESCRIPTION
**Pull request by Alhadis**
References #14279, which was inexplicably abandoned by @DomT4 after @ilovezfs had reviewed his revisions.

1. { 𝒳 } Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
2. { 𝒳 } Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
3. { 𝒳 } Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?  
	* Yes, but it installed `perl` to `HEAD-b9645564` or something similar. Which affected the version of `perl` that was picked up in my `$PATH`, causing `/usr/bin/perl` to be used instead.
	* I also installed from the tarball's source, and it built like absolute clockwork. Every single test passed.
4. { 𝒳 } Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?  
	* No, although it technically doesn't fail either:
~~~console
==> Installing or updating 'rubocop' gem
dyld: lazy symbol binding failed: Symbol not found: _rb_ary_new_from_values
  Referenced from: /Users/johngardner/.rvm/gems/ruby-2.2.2/gems/psych-2.2.4/lib/psych.bundle
  Expected in: flat namespace

dyld: Symbol not found: _rb_ary_new_from_values
  Referenced from: /Users/johngardner/.rvm/gems/ruby-2.2.2/gems/psych-2.2.4/lib/psych.bundle
  Expected in: flat namespace

Abort trap: 6
~~~

... I hate Ruby.

### CODE REVIEW

**Comment by ilovezfs** [comment]
@Alhadis thanks for picking this up :)

**Comment by Alhadis** [comment]
I love Perl. It's a tremendous honour.

... what's happening with CI?

**Comment by ilovezfs** [comment]
>... what's happening with CI?

what do you mean?

**Comment by DomT4** [comment]
> which was inexplicably abandoned by @DomT4

With respect, nobody is obligated to contribute to Homebrew and nobody is obligated to see a PR or Issue through to completion. 2708 commits to the core alone leaves me feeling fairly happy with how much of the workload I've taken on over the years.

**Comment by Alhadis** [comment]
> what do you mean?

Well, this ain't exactly filling me with fuzzy feelings:

Figure 1

> With respect, nobody is obligated to contribute to Homebrew and nobody is obligated to see a PR or Issue through to completion. 2708 commits to the core alone leaves me feeling fairly happy with how much of the workload I've taken on over the years.

Don't take that as a jab. I meant I was honestly expecting another rodeo of build difficulties here, which was my previous experience when we were dealing with v8. I was surprised to see the job done in a few minutes, which has given me the impression I might've missed something.

**Comment by DomT4** [comment]
> Don't take that as a jab.

Apologies. "Inexplicably" can be something of a loaded word. 

You'll want most of what I did in my PR, minus the style tweaks ILZ objected to, plus deciding what to do with `rex`, minus bothering to apply any fixes to the versioned `vim` that the current `vim` has since gained. Major `perl` releases are not backwards compatible, everything in `brew uses` needs bumped/fixing/etc.

**Comment by Alhadis** [comment]
Pretty confident I can handle Vim. Can't say as much for Rex:

~~~console
λ Formula (perl-5.26.0): brew info rex
rex: stable 1.5.0 (bottled)
Command-line tool which executes commands on remote servers.
https://rexify.org
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/rex.rb
==> Requirements
Required: perl >= 5.16 ✔
λ Formula (perl-5.26.0): 
~~~

I really hope testing this doesn't involve remote server access, because I no longer have access to any servers. :p Broke, jobless, no money for webhosting, et cetera.

**Comment by DomT4** [comment]
`vim` is nice and simple, usually it's just a revision bump. When it comes to `rex` either upstream will need to cut a new release to solve the issues ILZ objected to, or it needs to be shunted onto the system `perl`. The existing `test do` block on that formula was enough to let me know it was having various issues, so should hopefully do the same with you.

**Comment by ilovezfs** [comment]
The stakes are low
```
Josephs-MacBook-Pro:~ joe$ brew formula-analytics rex
install events in the last 30 days for rex
=====================================================================================================
1 | rex                                                                                | 23 | 100.00%
=====================================================================================================
Total                                                                                  | 23 |    100%
=====================================================================================================
```

**Comment by DomT4** [comment]
If upstream hasn't cut a new release with support yet, maybe just worth trashing the formula until they do then. Ideally it would be using the system `perl` but if it gets switched the hashbang situation needs to be fixed, which is fairly tedious as well.

**Comment by Alhadis** [comment]
Sorry, back.

What should I do about "official" taps like `homebrew/science` and co? In addition to the formulae updated in #14279, `brew uses` also brings up [`homebrew/science/libbi`](https://github.com/Homebrew/homebrew-science/blob/master/libbi.rb) and [`homebrew/science/links-scaffolder`](https://github.com/Homebrew/homebrew-science/blob/master/links-scaffolder.rb).

If we need to accommodate those dependants too, then I guess I'll need to be tapping every untapped repository to search...

**Comment by ilovezfs** [comment]
No need to worry about them.

**Comment by Alhadis** [comment]
Got it. Screw silence!

~~~console
curl: (22) The requested URL returned error: 404 Not Found
Error: Failed to download resource "rex"
Download failed: https://homebrew.bintray.com/bottles/rex-1.5.0_1.sierra.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Using the sandbox
==> Downloading https://cpan.metacpan.org/authors/id/J/JF/JFRIED/Rex-1.5.0.tar.gz
######################################################################## 100.0%
==> Downloading http://search.cpan.org/CPAN/authors/id/L/LE/LEONT/Module-Build-0.4222.tar.gz
==> Downloading from http://www.cpan.org/authors/id/L/LE/LEONT/Module-Build-0.4222.tar.gz
######################################################################## 100.0%
==> perl Makefile.PL INSTALL_BASE=/usr/local/Cellar/rex/1.5.0_1/libexec
Last 15 lines from /Users/johngardner/Library/Logs/Homebrew/rex/01.perl:
2017-06-23 02:58:52 +1000

perl
Makefile.PL
INSTALL_BASE=/usr/local/Cellar/rex/1.5.0_1/libexec

`vxs.c`: loadable library and perl binaries are mismatched (got handshake key 0xd880080, needed 0xdb00080)

READ THIS: http://docs.brew.sh/Troubleshooting.html
~~~

File: Formula/vim.rb
Comment (JCount) [inline Formula/vim.rb]: This revision line needs to go before the `head` line.

**Review by JCount — DISMISSED** [review]
@Alhadis Could you please split this up so that you have one commit for each changed file. Also, the perl alias needs to be updated. An easy way to do this is, from the root of the `homebrew/core` repo, run `git mv Aliases/perl\@5.24 Aliases/perl\@5.26` and then add it in it's own commit.

**Comment by Alhadis** [comment]
@JCount
> Could you please split this up so that you have one commit for each changed file.

Are you saying I should rebase e835e7 or something...? :\

**Comment by Alhadis** [comment]
@JCount I'm getting this error when I run `brew audit --strict perl`:

~~~
perl:
  * Formula has invalid versioned aliases:
      perl@5.24
Error: 1 problem in 1 formula
~~~

It was previously giving me two until I fixed this one:
~~~
Formula has other versions so create a versioned alias:
      cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/perl.rb perl@5.26
~~~

Now when I run your recommended `git mv Aliases/perl\@5.24 Aliases/perl\@5.26` line, I get:

~~~
fatal: destination exists, source=Aliases/perl@5.24, destination=Aliases/perl@5.26
~~~

Absolutely no idea what's going on here... :\

File: Formula/vim.rb
Comment (Alhadis) [inline Formula/vim.rb]: Done.

**Comment by JCount** [comment]
@Alhadis Have you ever done an interactive git rebase? http://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request.html provides some information.

**Comment by Alhadis** [comment]
Alrighty, I dropped both commits related to revision-bumping, and redid them one-by-one for each formula. =)

**Comment by JCount** [comment]
Thanks @Alhadis 😃

**Comment by Alhadis** [comment]
Green at last! =)

File: Formula/rex.rb
Comment (ilovezfs) [inline Formula/rex.rb]: This needs to be reported upstream with a link to the report here.

File: Formula/rex.rb
Comment (ilovezfs) [inline Formula/rex.rb]: I don't think we need this since users can install arbitrary versions of perl on lower macOS versions.

File: Formula/rex.rb
Comment (ilovezfs) [inline Formula/rex.rb]: These should be using the https cpan.metacpan.org links.

File: Formula/rex.rb
Comment (Alhadis) [inline Formula/rex.rb]: @ilovezfs Did you [see the comment](https://github.com/Homebrew/homebrew-core/pull/14279/files#r123865541) I left on @DomT4's PR? There's nothing to report. The author of the JSON::XS module (who isn't the author of `rex`) is an evangelical idiot with some vendetta against the new Perl 5 maintainers. He's not going to "fix" this because he's **choosing** not to support new Perl versions.

File: Formula/rex.rb
Comment (Alhadis) [inline Formula/rex.rb]: Force-pushed with an amended URL, and `depends_on` removed.

File: Formula/rex.rb
Comment (ilovezfs) [inline Formula/rex.rb]: Yes, and this needs to be reported to `rex` upstream regardless.

Comment (Alhadis) [inline Formula/rex.rb]: His FAQ actually states that if your build or test-suite passes, there's nothing to worry about:

> **"Does this mean the module will not work?"**  
> "No, in most cases, it will work. Currently only Coro will definitely not work with 5.22.0, while most or even all of my other modules will probably still work.  
>  
> **So as long as everything seems to work (and/or the testsuite runs successfully), you are most likely in business and can ignore the warning.**

And then there's this:
> What it does mean, and I am quite serious about this, is that I don't want to hear about problems with perl 5.22.0 or any later release - these are simply not supported. They might work perfectly fine, but you are definitely on your own.

Comment (ilovezfs) [inline Formula/rex.rb]: Right. To `rex` upstream, not this guy. `rex` is using a module that is causing issues for downstream and that needs to be reported and linked here.

Comment (Alhadis) [inline Formula/rex.rb]: The only issue is a petty, preachy warning during a build process. You honestly expect them to revise their project's dependencies over this? Hell, they'd probably ignore it because `JSON::XS` has a lot going for it in terms of speed and performance. Not worth giving up over a non-issue.

Comment (ilovezfs) [inline Formula/rex.rb]: >You honestly expect them to revise their project's dependencies over this?

Yes. Using a module not supported on current perl, and expecting every downstream to set an environment variable to override that, in perpetuity, isn't acceptable.

Comment (Alhadis) [inline Formula/rex.rb]: Sure. What if they refuse? Are you telling me newer Perl releases will be blocked forever because of some cranky, idiot developer's prejudices?

Comment (ilovezfs) [inline Formula/rex.rb]: That's entirely up to them, but would be sufficient reason to remove rex from Homebrew.

Comment (Alhadis) [inline Formula/rex.rb]: Oh right, `rex`. Sorry, somehow shifted focus in my mind and thought this was blocking the Perl formula. Spent way too long reading that dude's ranting...

Comment (ilovezfs) [inline Formula/rex.rb]: Yeah, rex. I don't think there's any hope of getting the module author to change his opinions, but I do think we can reasonably expect rex to say sure we'll change the dep to something supported, or no we'd rather not change the dep despite the downstream consequences. Either way, that will provide a rational basis for either keeping or removing rex.

But perl will be upgraded regardless, and, no, we do not need to wait for them to respond.

Comment (Alhadis) [inline Formula/rex.rb]: Done. Reported in RexOps/Rex#1153.

Merge conflict also resolved.

File: Formula/rex.rb
Comment (ilovezfs) [inline Formula/rex.rb]: Thanks!

**Review by ilovezfs — DISMISSED** [review]
Please use https://cpan.metacpan.org/... for the primary URLs of the perl modules so that they're using https.

**Comment by Alhadis** [comment]
Remaining HTTP links replaced. You guys should probably have this sort of thing checked by `brew audit`... at least for known vendors like CPAN.

Also, [Jenkins crapped his pants](https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/3555/version=sierra/testReport/junit/brew-test-bot/sierra/pull___clean_https___github_com_Homebrew_homebrew_core_pull_14849/) the moment I pushed, and in no time at all. Impressive.

~~~
Error Message

failed: brew pull --clean https://github.com/Homebrew/homebrew-core/pull/14849
Stacktrace

        fatal: ref HEAD is not a symbolic ref
==> Fetching patch 
Patch: https://github.com/Homebrew/homebrew-core/pull/14849.patch
==> Applying patch
Applying: perl 5.26.0
Applying: perl: Replace perl@5.24 alias with perl@5.26
Applying: rex: bump revision for perl
Applying: subversion: bump revision for perl
Applying: vim: bump revision for perl
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	Formula/vim.rb
Falling back to patching base and 3-way merge...
Auto-merging Formula/vim.rb
CONFLICT (content): Merge conflict in Formula/vim.rb
Patch failed at 0005 vim: bump revision for perl
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
Error: Patch failed to apply: aborted.
~~~

**Comment by ilovezfs** [comment]
I will fix it for you.

**Comment by Alhadis** [comment]
Just let me know if there's anything more I can do. =)

**Comment by ilovezfs** [comment]
Hopefully we'll have :green_apple: now!

**Comment by Alhadis** [comment]
> Hopefully we'll have 🍏 now!

Good! I love apples. Totally kick the shit out of oranges.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
@Alhadis Thanks for your first contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**
